### ISSUE TITLE
Why fit is not merged into fit_generator since the latter is more generalized?

### ISSUE THREAD

**Issue by ghostplant** [keras-team/keras#9667]
Here are the method of `fit` and `fit_generator`:

```sh
    def fit(self,
            x=None,
            y=None,
            batch_size=None,
            epochs=1,
            verbose=1,
            callbacks=None,
            validation_split=0.,
            validation_data=None,
            shuffle=True,
            class_weight=None,
            sample_weight=None,
            initial_epoch=0,
            steps_per_epoch=None,
            validation_steps=None,
            **kwargs):
```

```sh
    def fit_generator(self,
                      generator,
                      steps_per_epoch=None,
                      epochs=1,
                      verbose=1,
                      callbacks=None,
                      validation_data=None,
                      validation_steps=None,
                      class_weight=None,
                      max_queue_size=10,
                      workers=1,
                      use_multiprocessing=False,
                      shuffle=True,
                      initial_epoch=0):
```

The interface is almost the same while they are totally implemented separately, why not wrap `fit` with `fit_generator + IndexArrayGenerator`? Similarly, there are also `evaluate_generator`, etc.

**Comment by daverio** [comment]
I completely agree with you :-) 
I would do a fit2 which is a function which is called by fit and fit_generator. But where it should be done? In Keras or Keras-contrib?

**Comment by ghostplant** [comment]
`fit_generator` lacks sample_weight which should be solved in advanced, I can help with this.

**Comment by daverio** [comment]
It is not only sample_weight which need to be added.

I think fit_generator should also call the _fit_loop(). But it means that the _fit_loop has to be able to works with sequence/generator and not only with arrays...

I would say:
Create a _fit_loop2, fit2, _test_loop2, evaluate2, _predict_loop2, predict2.
Once those works well and are tested, wrap X and X_generator to them.

But all this means a bit of work and a massive redesign of training.py, the core of Keras...  From the guideline, I think one should make a proposition and see if it is accepted.

I really cant bother to write this document, so 
@ghostplant would you do it?

But if accepted, I can give 2 days of work to do it :-).

**Comment by ghostplant** [comment]
I dont think this method is suitable since large refactor will bring many new bugs. fit_generator needs to be enhanced in my opinion.

**Comment by daverio** [comment]
What fit_generator is lacking of? for now I see only the missing sample_weight, then that maybe one could add x and y as parameters (and check if either the generator is passed or x and y). Then you would mimic completely the fit function. But then merging both fit and fit_generator would need to think about _fit_loop.
The idea of externalizing logic is a good one, and I expect autors to wish to keep it :-).

**Comment by ghostplant** [comment]
https://github.com/keras-team/keras/pull/9680/files

@daverio The only thing is that `fit_generator` need a case check to support framework-native tensor data, after that, legacy `_fit_loop` can be fully dropped.

**Comment by tRosenflanz** [comment]
Plus fit_generator does support sample weights through the generator itself: you can output a tuple of (data, target, sample_weight)

**Comment by ghostplant** [comment]
@tRosenflanz Thanks, it seems `fit` has some other special usages, such as dict-like same_weight, etc.

**Comment by ghostplant** [comment]
All differences listed below (I just skip their common points that are originally supported by both):

Status | legacy_loop of Fit/Evaluate/Predict | generator of Fit/Evaluate/Predict
------------ | ------------ | -------------
Dynamic Inputs | Cannot Support | Originally Support
Multi-thread Perform | Cannot Support | Originally Support
Full Verbose Mode | Supported | Supported after #9811
Train/Test on Native Tensors | Supported | Supported after #9816
Validation on Native Tensors | Supported | Not Supported // TODO
Sparse Input Array | Supported | Not Supported // TODO
Input Array from HDF5 | Supported | Not Supported // TODO

After resolving all issues above, they can be merged eventually. I'll continue to future updates on this path.
